### ISSUE TITLE
Is is possible to mock Motor with mongomock for unit testing purposes?

### ISSUE THREAD

**Issue by y3g0r** [mongomock/mongomock#244]
I know that mongomock is used to mock pymongo client.
I'm using tornado framework and communicate with Mongo using [Motor](https://motor.readthedocs.io/en/stable/).
Motor is using pymongo as its underlying mongo client (see [MongoClient](https://github.com/mongodb/motor/blob/master/motor/motor_tornado.py#L33) and [AgnosticClient](https://github.com/mongodb/motor/blob/876662955ba3aae1355fb2217be4d129553ac50c/motor/core.py#L497))
Motor [is using greenlet](https://emptysqua.re/blog/motor-internals-how-i-asynchronized-a-synchronous-library/) to make synchronous pymongo client work in asynchronous fashion. I don't really know how they do it (I'm not expert in greenlet or whatever).

My question is can I use mongomock to mock Motor client?
So basically I need mongomock functionality, but my application is expecting to get Futures as response to all communications with Mongo.
Maybe I don't have to use Motor but I can wrap mongomock to my custom class, which will provide async behavior that my client code is expecting from mongo client.

Thank you

**Comment by drorasaf** [comment]
you can consider doing something like:
from mock import MagicMock
import mongomock
p = mock.patch('motor.MotorClient.__delegate_class__', new=mongomock.MongoClient)
p1 = mock.patch('motor.MotorDatabase.__delegate_class__', new=MagicMock)

p.start()
p1.start()

def fin():
    p.stop()
    p1.stop()

request.addfinalizer(fin)

**Comment by pcorpet** [comment]
Hi @y3g0r , sorry for the late answer. I've recently added a `patch` function to mongomock that could probably help with Motor. I've never used Motor itself, but calling `mongomock.patch` as a decorator on a test class or function, replaces calls to `pymongo.MongoClient` by calls to mongomock transparently.

Can you test it and tell us if it works?

**Comment by elatt** [comment]
Unfortunately, I don't think your suggestion @pcorpet quite works. Here are a few issues:
* The test function I want to use the mock in is `async` so using `mongomock.patch` as a decorator doesn't seem possible. However, I believe this is an easy workaround as I can just use it as a context-manager
* `Motor` seems to want to import `pymongo.mongo_client.MongoClient` so the mongomock.patch isn't picking it up
* Finally, even if those things are resolved, `Motor` also seems to want to subclass `MongoClient` so if we patch it as a function, we have `super()` calls that fail.

**Comment by 3mp3ri0r** [comment]
Any update about implementing mongomock to motor?

**Comment by pcorpet** [comment]
I think that @elatt is right. I went to take a deeper look at `Motor` and don't see any easy way to mock it with `mongomock`.
* The implementation of `Motor` is deeply linked to the one of `pymongo`
* `Motor` is counting on `socket` to block calls, but `mongomock` never blocks anything as everything is synchronous. Moreover, `mongomock` is built on the assumption that all calls to it should be blocking.

So I think it's non-trivial to use `mongomock` for unit testing code using `Motor`.

If one of you has a good idea, or at least a hint, I would welcome any suggestions.

**Comment by elbehery95** [comment]
What if mongomock provides some factory that returns an async wrapper for pymongo ? would this be enough for testing async based logic ?

E.g
```python
@mongomock.patch(servers=(('server.example.com', 27017),))
@pytest.mark.asyncio
async def test_increate_votes_endpoint():
  objects = [dict(votes=1), dict(votes=2), ...]
  client = mongomock.to_async(pymongo.MongoClient('server.example.com'))
  await client.db.collection.insert_many(objects)
  # rest of checking logic
```

**Comment by Goldziher** [comment]
this would be great @elbehery95

**Comment by kyleposluns** [comment]
Any plans on adding support for this?

**Comment by elbehery95** [comment]
@kyleposluns  I was trying to add this in https://github.com/mongomock/mongomock/pull/767 using proxy pattern. But PR was not active

**Comment by kyleposluns** [comment]
@elbehery95  Well, I would still love to see this added at some point.

**Comment by kyleposluns** [comment]
I am currently using proxy pattern to get this to work, but I'd still like this to be supported in mongomock please!

**Comment by ramnes** [comment]
I think y'all might want to take a look at [mongomock-motor](https://github.com/michaelkryukov/mongomock_motor). :)
